client/components/app.jsx

```diff
@@ -0,0 +1,136 @@
+import React, { Component } from 'react';
+import Header from './layout/Header';
+
+export default class TipCalculator extends Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      tipPercentage: {
+        '15%': 0.15,
+        '16%': 0.16,
+        '17%': 0.17,
+        '18%': 0.18,
+        '19%': 0.19,
+        '20%': 0.20,
+      },
+      tipSelected: null,
+      checkTotal: '',
+      checkTotalSet: false,
+      tipTotal: null,
+      grandTotal: null,
+      fateDecides: false,
+    };
+    this.onTotalCostChange = this.onTotalCostChange.bind(this);
+    this.onSetClick = this.onSetClick.bind(this);
+    this.calculateGrandTotal = this.calculateGrandtotal.bind(this);
+    this.handleTipPercentageClick = this.handleTipPercentageClick.bind(this);
+    this.handleReset = this.handleReset.bind(this);
+    this.letFateDecide = this.letFateDecide.bind(this);
+  }
+
+  onTotalCostChange(e) {
+    const name = e.target.name, value = e.target.value;
+    this.setState({ [name]: value });
+  }
+
+  onSetClick() {
+    this.setState({ checkTotalSet: !this.state.checkTotalSet }, () => {
+      if (this.state.fateDecides) {
+        const tipArr = this.state.tipSelected.split('');
+        tipArr.pop();
+        const tip = '0.' + tipArr.join();
+        this.calculateGrandTotal(tip)
+      }
+      else this.calculateGrandTotal(this.state.tipSelected)
+    })
+  }
+
+  handleReset() {
+    this.setState({ checkTotalSet: !this.state.checkTotalSet, tipTotal: null, grandTotal: null, checkTotal: '', tipSelected: null, fateDecides: false })
+  }
+
+  handleTipPercentageClick(e) {
+    if (this.state.fateDecides) {
+      const randomTip = Math.round(Math.random() * (40 - 15) + 15);
+      const percentage = '0.' + randomTip;
+      const tipSelected = percentage.split('.')[1] + '%';
+      this.setState({ tipSelected }, () => this.calculateGrandTotal(percentage))
+    } else {
+      const tipSelected = e.target.value;
+      this.setState({ tipSelected }, () => this.calculateGrandtotal(tipSelected));
+    }
+  }
+
+  letFateDecide(e) {
```

Comment:
Funny name for the display text on the button, but try a function name with a stronger semantic value like randomizeTipPercentage.